examples/react/kitchen-sink-file-based/src/components/Breadcrumbs.tsx

```diff
@@ -0,0 +1,28 @@
+import { isMatch, Link, useMatches } from '@tanstack/react-router'
+
+export const Breadcrumbs = () => {
+  const matches = useMatches()
+
+  if (matches.some((match) => match.status === 'pending')) return null
+
+  const matchesWithCrumbs = matches.filter((match) =>
+    isMatch(match, 'loaderData.crumb'),
+  )
```

Comment:
looking at [the sandbox](https://stackblitz.com/edit/dj9kzo?file=src%2Fcomponents%2FBreadcrumbs.tsx), `matchesWithCrumbs` is always `RouteMatch<any, any, any, any, any, any, any, any>[]`:

![Screenshot 2024-08-02 at 17 01 52](https://github.com/user-attachments/assets/adb7f2ff-78ad-4bce-bfe2-fa54f2322d33)

Is that because the sandbox is using the wrong TS version?

Reply:
also - why couldn't we just check for `match.loaderData?.crumb !== undefined` ?

Reply:
I noticed this. It's weird. Works well in vscode and sometimes in the sandbox

Reply:
> also - why couldn't we just check for `match.loaderData?.crumb !== undefined` ?

The issue is TS is not able to narrow this. If you have loader data in matches which do not have `crumb`, you can't get to the property without using `in`. Furthermore TS is not able to narrow the parent type of a property if it's not a literal type. This is why `routeId` works but `crumb` does not. Discriminated unions work but other properties do not

Reply:
> I noticed this. It's weird. Works well in vscode and sometimes it works in the sandbox but not all the time

Maybe the types just need optimising more. I think there's still too much instantiations happening in the recursive type due to it distributing over the recursive type. Maybe stackblitz is more limited in resources or uses an older version of TS which kicks the bucket with more complex types?